### PR TITLE
build: Adds Windows kube-proxy image

### PR DESCRIPTION
**Pull request by claudiubelu**
#### What type of PR is this?

/sig windows

/kind feature

#### What this PR does / why we need it:

Adds the ``KUBE_BUILD_WINDOWS`` option to ``make release-images`` and ``quick-release-images``, which will allow it to build the a Windows kube-proxy image as well. That image can then be used with Windows Host Process Containers to start the kube-proxy service on Windows nodes.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Added support for building Windows kube-proxy container image.
A container image for kube-proxy on Windows can now be built with the command
`make release-images KUBE_BUILD_WINDOWS=y`.
The Windows kube-proxy image can be used with Windows Host Process Containers.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

File: build/server-image/kube-proxy/Dockerfile_windows [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Dockerfile used for the server images.
+
+# Note that the image doesn't really matter for Windows Host Process containers
+# # the files in the image are copied to $env:CONTAINER_SANDBOX_MOUNT_POINT on the host
```
Comment (jsturtevant) [inline build/server-image/kube-proxy/Dockerfile_windows:18]: in containerd 1.7+ the files will be mounted is `c:/hpc`

**Comment by claudiubelu** [comment]
Logs from the latest build: https://paste.ubuntu.com/p/nB55kXb56y/

Pushed as ``docker.io/claudiubelu/kube-proxy-windows-amd64:pr-109939-test`` with crane:

```
crane push _output/release-images/amd64/kube-proxy.exe.tar  docker.io/claudiubelu/kube-proxy-windows-amd64:pr-109939-test
2024/03/05 22:50:19 existing blob: sha256:2520574c99c083ab00531d22b89eb9ee9d38a470b82bbd5f69169f1770dffcbb
2024/03/05 22:50:20 pushed blob: sha256:865f4dd92e9a2aee8405af18b553f0aa3d33fffb0f71f137afb981d85c36a33b
2024/03/05 22:50:21 pushed blob: sha256:69b4843bac8c03b2b826b8bd2a472ec99e9fb60ca3cf81ee2b96d32d0d18f2e2
2024/03/05 22:50:22 docker.io/claudiubelu/kube-proxy-windows-amd64:pr-109939-test: digest: sha256:a654ea1789553467dcc679ce03680d6c4b8e8dbde6109c8be635372d6749b7f0 size: 589
index.docker.io/claudiubelu/kube-proxy-windows-amd64@sha256:a654ea1789553467dcc679ce03680d6c4b8e8dbde6109c8be635372d6749b7f0
```

File: build/common.sh [hunk 2]
```diff
@@ -107,28 +107,35 @@ readonly KUBE_APISERVER_BASE_IMAGE="${KUBE_APISERVER_BASE_IMAGE:-$KUBE_GORUNNER_
...
 readonly KUBE_SCHEDULER_BASE_IMAGE="${KUBE_SCHEDULER_BASE_IMAGE:-$KUBE_GORUNNER_IMAGE}"
 readonly KUBE_PROXY_BASE_IMAGE="${KUBE_PROXY_BASE_IMAGE:-$KUBE_BASE_IMAGE_REGISTRY/distroless-iptables:$__default_distroless_iptables_version}"
+readonly KUBE_PROXY_WINDOWS_BASE_IMAGE="${KUBE_PROXY_WINDOWS_BASE_IMAGE:-mcr.microsoft.com/oss/kubernetes/windows-host-process-containers-base-image:v0.1.0}"
 readonly KUBECTL_BASE_IMAGE="${KUBECTL_BASE_IMAGE:-$KUBE_GORUNNER_IMAGE}"
 
 # This is the image used in a multi-stage build to apply capabilities to Docker-wrapped binaries.
 readonly KUBE_BUILD_SETCAP_IMAGE="${KUBE_BUILD_SETCAP_IMAGE:-$KUBE_BASE_IMAGE_REGISTRY/setcap:$__default_setcap_version}"
 
-# Get the set of master binaries that run in Docker (on Linux)
+# Get the set of master binaries that run in Docker (on Linux), or as Host Processes on Windows.
```
Comment (jsturtevant) [inline build/common.sh:116]:
```suggestion
# Get the set of master binaries that run in Docker (on Linux), or as Host Processes Container on Windows.
```

File: build/common.sh [hunk 3]
```diff
@@ -107,28 +107,35 @@ readonly KUBE_APISERVER_BASE_IMAGE="${KUBE_APISERVER_BASE_IMAGE:-$KUBE_GORUNNER_
...
 # Entry format is "<binary-name>,<base-image>".
-# Binaries are placed in /usr/local/bin inside the image.
+# Binaries are placed in /usr/local/bin inside the image (Linux), or C:\bin (Windows).
```
Comment (jsturtevant) [inline build/common.sh:118]:
```suggestion
# Binaries are placed in /usr/local/bin inside the image (Linux), or c:\hpc (Windows).
```

File: build/lib/release.sh [hunk 4]
```diff
@@ -282,17 +298,20 @@ function kube::release::build_conformance_image() {
 # This builds all the release docker images (One docker image per binary)
 # Args:
 #  $1 - binary_dir, the directory to save the tared images to.
-#  $2 - arch, architecture for which we are building docker images.
+#  $2 - os, the OS for which we are building docker images.
```
Comment (jsturtevant) [inline build/lib/release.sh:301]: should OS be added to the end of this so it can be optional and not break existing tooling?

**Review by jsturtevant — COMMENTED** [review]
couple minor comments

/lgtm

**Comment by jsturtevant** [comment]
looks like there is a liniting issue:

```
          -t "${docker_image_tag}" ${additional_tag} \
                                   ^---------------^ SC2086 (info): Double quote to prevent globbing and word splitting.
```

**Comment by sftim** [comment]
Changelog suggestion
```diff
-Windows kube-proxy can now be built with the command ``make release-images KUBE_BUILD_WINDOWS=y``. The Windows kube-proxy image can be used with Windows Host Process Containers.
+Added support for building Windows kube-proxy container image.
+A container image for kube-proxy on Windows can now be built with the command
+`make release-images KUBE_BUILD_WINDOWS=y`.
+The Windows kube-proxy image can be used with Windows Host Process Containers.
```

**Comment by claudiubelu** [comment]
> Changelog suggestion
> 
> ```diff
> -Windows kube-proxy can now be built with the command ``make release-images KUBE_BUILD_WINDOWS=y``. The Windows kube-proxy image can be used with Windows Host Process Containers.
> +Added support for building Windows kube-proxy container image.
> +A container image for kube-proxy on Windows can now be built with the command
> +`make release-images KUBE_BUILD_WINDOWS=y`.
> +The Windows kube-proxy image can be used with Windows Host Process Containers.
> ```

Done

File: build/common.sh (hunk 3 above)
Comment (jsturtevant) [reply build/common.sh]: > Do you suggest we just copy the kube-proxy binary and other scripts to `C:\` in the Dockerfile?

No I don't think so.  I think its just a bit confusing... If you start a HPC pod it won't be at `c:\bin` but at `c:\hpc\bin`, maybe there is a way to word it so its clearer but I am not sure.   If I read the dockerfile correctly its actually at `c:/hpc/kube-proxy/kube-proxy.exe`?

**Comment by claudiubelu** [comment]
Latest logs with windows image: https://paste.ubuntu.com/p/cdwM2Jnp5d/

**Comment by claudiubelu** [comment]
Can we get this in in this cycle?

**Review by jsturtevant — COMMENTED** [review]
/lgtm

**Comment by dims** [comment]
@jsturtevant we need at least one person from release related areas to chime in. let's try @jeremyrickard ?

File: build/common.sh [hunk 5]
```diff
@@ -107,28 +107,35 @@ readonly KUBE_APISERVER_BASE_IMAGE="${KUBE_APISERVER_BASE_IMAGE:-$KUBE_GORUNNER_
 readonly KUBE_CONTROLLER_MANAGER_BASE_IMAGE="${KUBE_CONTROLLER_MANAGER_BASE_IMAGE:-$KUBE_GORUNNER_IMAGE}"
 readonly KUBE_SCHEDULER_BASE_IMAGE="${KUBE_SCHEDULER_BASE_IMAGE:-$KUBE_GORUNNER_IMAGE}"
 readonly KUBE_PROXY_BASE_IMAGE="${KUBE_PROXY_BASE_IMAGE:-$KUBE_BASE_IMAGE_REGISTRY/distroless-iptables:$__default_distroless_iptables_version}"
+readonly KUBE_PROXY_WINDOWS_BASE_IMAGE="${KUBE_PROXY_WINDOWS_BASE_IMAGE:-mcr.microsoft.com/oss/kubernetes/windows-host-process-containers-base-image:v0.1.0}"
```
Comment (marosset) [inline build/common.sh:110]: let's use v1.0.0

**Comment by marosset** [comment]
After reviewing the code and testing this locally everything mostly looks good.
Let's update the base image ref tho

**Comment by jeremyrickard** [comment]
address the comment from Mark and lgtm

File: build/common.sh (hunk 5 above)
Comment (claudiubelu) [reply build/common.sh]: Done

**Comment by claudiubelu** [comment]
> address the comment from Mark and lgtm

Done

**Comment by dims** [comment]
/approve
/lgtm

thanks @claudiubelu @marosset @jeremyrickard

File: build/common.sh [hunk 6]
```diff
@@ -107,28 +107,35 @@ readonly KUBE_APISERVER_BASE_IMAGE="${KUBE_APISERVER_BASE_IMAGE:-$KUBE_GORUNNER_
 readonly KUBE_CONTROLLER_MANAGER_BASE_IMAGE="${KUBE_CONTROLLER_MANAGER_BASE_IMAGE:-$KUBE_GORUNNER_IMAGE}"
 readonly KUBE_SCHEDULER_BASE_IMAGE="${KUBE_SCHEDULER_BASE_IMAGE:-$KUBE_GORUNNER_IMAGE}"
 readonly KUBE_PROXY_BASE_IMAGE="${KUBE_PROXY_BASE_IMAGE:-$KUBE_BASE_IMAGE_REGISTRY/distroless-iptables:$__default_distroless_iptables_version}"
+readonly KUBE_PROXY_WINDOWS_BASE_IMAGE="${KUBE_PROXY_WINDOWS_BASE_IMAGE:-mcr.microsoft.com/oss/kubernetes/windows-host-process-containers-base-image:v1.0.0}"
```
Comment (liggitt) [inline build/common.sh:110]: shouldn't this default to a kubernetes-project-controlled registry?

File: build/lib/release.sh [hunk 7]
```diff
@@ -173,14 +174,24 @@ function kube::release::package_node_tarballs() {
 # Package up all of the server binaries in docker images
 function kube::release::build_server_images() {
   kube::util::ensure-docker-buildx
+  docker buildx create --name img-builder --use || true
+  docker buildx inspect --bootstrap
+  kube::util::trap_add 'docker buildx rm' EXIT
+
+  platforms=("${KUBE_SERVER_PLATFORMS[@]}")
+  if [[ "${KUBE_BUILD_WINDOWS}" =~ [yY] ]]; then
+    platforms=("${KUBE_SERVER_PLATFORMS[@]}" "windows/amd64")
```
Comment (liggitt) [inline build/lib/release.sh:183]: this override is *really* hard to follow ... it makes KUBE_SERVER_PLATFORMS non-authoritative for `kube::release::build_server_images`, and relies on a bash variable not defined in this file ... 

why override this way instead of invoking with KUBE_SERVER_PLATFORMS=("windows/amd64") or something similar?

File: build/release-images.sh [hunk 8]
```diff
@@ -39,6 +39,10 @@ kube::build::verify_prereqs
 kube::build::build_image
 kube::build::run_build_command make all WHAT="${CMD_TARGETS}" KUBE_BUILD_PLATFORMS="${KUBE_SERVER_PLATFORMS[*]}" DBG="${DBG:-}"
 
+if [[ "${KUBE_BUILD_WINDOWS}" =~ [yY] ]]; then
```
Comment (liggitt) [inline build/release-images.sh:42]: does using a bash variable not defined in this script without a default break `make release-images`?

File: build/release-images.sh [hunk 9]
```diff
@@ -39,6 +39,10 @@ kube::build::verify_prereqs
...
+  kube::build::run_build_command make all WHAT="cmd/kube-proxy" KUBE_BUILD_PLATFORMS="windows/amd64"
```
Comment (liggitt) [inline build/release-images.sh:43]: hard-coding specific binaries and platforms into this script doesn't seem correct ... shouldn't these be fed off  CMD_TARGETS and KUBE_SERVER_PLATFORMS?

File: hack/lib/golang.sh [hunk 10]
```diff
@@ -101,7 +101,8 @@ kube::golang::server_image_targets() {
 
 IFS=" " read -ra KUBE_SERVER_IMAGE_TARGETS <<< "$(kube::golang::server_image_targets)"
 readonly KUBE_SERVER_IMAGE_TARGETS
-readonly KUBE_SERVER_IMAGE_BINARIES=("${KUBE_SERVER_IMAGE_TARGETS[@]##*/}")
+readonly KUBE_SERVER_LINUX_IMAGE_BINARIES=("${KUBE_SERVER_IMAGE_TARGETS[@]##*/}")
+readonly KUBE_SERVER_WINDOWS_IMAGE_BINARIES=("kube-proxy.exe")
```
Comment (liggitt) [inline hack/lib/golang.sh:105]: hard-coding specific binaries and platforms into this script doesn't seem correct ... shouldn't these be defined alongside KUBE_SERVER_IMAGE_TARGETS, not here?

Comment (liggitt) [reply hack/lib/golang.sh:105]: also, can we leave KUBE_SERVER_IMAGE_BINARIES as-is and add a KUBE_SERVER_WINDOWS_IMAGE_BINARIES, rather than renaming the existing variable? it's a lot harder to reason about impact of changes that are not strictly additive

**Comment by liggitt** [comment]
sorry for the post-merge review, the renames of existing server target variables broke some downstream test / build invocations ... is there a way to make this strictly additive and drive the windows build from the top-level invocation rather than hard-coding windows cases into multiple files?

**Comment by aojea** [comment]
@claudiubelu is it problematic to revert and reimplement this with Jordan's suggestions?
@soltysh can you check if Openshift is impacted too ? I imagine more downstream consumers will be affected by this change as soon as they start consuming it

File: build/common.sh (hunk 6 above)
Comment (aojea) [reply build/common.sh:110]: can we move a copy of this image to our registry?
cc: @BenTheElder @ameukam

**Comment by liggitt** [comment]
The downstream stuff is secondary, that's just what made me notice it. I'm more concerned about keeping k/k build scripts from becoming more tangled. 

For example, this starts building windows server images but doesn't add windows to the "supported server platforms" list, and instead adds parallel special-cases throughout the build scripts.

I'd be happy to give tips about defining stuff in sensible ways from the top level vars, with helper functions to derive file names or whatever to account for windows .exe extensions, etc, I just won't be around until Thursday.

**Comment by jeremyrickard** [comment]
I’ve opened a revert PR if that would be ok with folks.

**Review by claudiubelu — COMMENTED** [review]
I've resent this PR + changes suggested here: https://github.com/kubernetes/kubernetes/pull/125297

File: build/common.sh (hunk 6 above)
Comment (BenTheElder) [reply build/common.sh:110]: heh https://github.com/kubernetes/kubernetes/pull/125297#discussion_r1624818540
